### ISSUE TITLE
Unused named function parameters cause parse error

### ISSUE THREAD

**Issue by davidfstr** [idris-lang/Idris-dev#1940]
This program compiles:

```
import Data.Vect
foo : Ord e => (Vect n e)
```

However this one does not:

```
import Data.Vect
foo : Ord e => (xs:Vect n e)
```

Error:

```
./InsertionSort_NoProofs.idr:2:29: error: unexpected
    EOF, expected: "->"
foo : Ord e => (xs:Vect n e)<EOF> 
```

Adding a name to a parameter of a function signature should not cause a parse error, even if the name is not subsequently used.

**Comment by davidfstr** [comment]
Environment: Idris 0.9.16 / OS X 10.9.5

**Comment by jfdm** [comment]
@davidfstr The reason that compilation fails is that return types are not function parameters and cannot be annotated (annotated might not be the best word here to use) as you expect.

Your first program compiles as it declares a `foo` with type `Vect n e` in which `e`'s are orderable.
Your second program on the other hand fails to compile as the compiler believes that you are still in the middle of writing a type signature and is expecting to see a return type.

**Comment by davidfstr** [comment]
It might still be useful to permit return types to be annotated, or at least recognize the case and give a nice error.

To provide a bit more context, I performed what I thought was a straightforward syntax transformation and got a parse error of this variety, which I found rather unexpected. To be specific, I started with:

```
insertionSort :
    Ord e =>
    (xs:Vect n e) ->
    (xs':Vect n e ** IsSorted xs')
```

And stripped out all the proof terms, leaving just:

```
insertionSort' :
    Ord e =>
    (xs:Vect n e) ->
    (xs':Vect n e)
```

Boom! Parse error. I consider this to be unexpected behavior.

Again I will reiterate: It might still be useful to permit return types to be annotated to deal with cases like this, or at least recognize the case and give a nice error.

**Comment by edwinb** [comment]
I think it should be a parse error. Names are often required on 
arguments, because they are really bindings - the name bound in an 
argument is in scope in the return type. On the return type, on the 
other hand, the name is meaningless (and I can't think of any other 
language where this sort of thing is supported, though perhaps such a 
language does exist).

I agree the error message should be better, but that is true of almost 
every parse error we get and is a rather bigger issue, probably 
requiring a significant rewrite of the parser unfortunately.

**Comment by davidfstr** [comment]
+1 for better parser error message.

> On the return type, on the other hand, the name is meaningless (and I can't think of any other language where this sort of thing is supported, though perhaps such a language does exist).

In this case I'm appealing to the desire for syntax that looks similar (i.e. everything separated by `->`) to actually be treated similarly, without special rules (such as the term after the last `->` being restricted).

**Comment by david-christiansen** [comment]
I don't think that things that are not similar should look similar, and the RHS of a dependent function type is not, in my opinion, particularly similar to what comes to the left of the arrow. This is because the arrow is a binder, just like lambda, let, and friends. If we were to follow the idea that these were, in fact, improved by naming the parts, then we should do it consistently, again to treat similar things as similar.

So `let` could allow syntax like `let x : Nat = 42 in y = x + x`, or `\x : Nat => y : x + x`, which would fit in with `(x : Nat) -> (y : Vect x a)` being allowed. But I think that these kinds of changes are more likely to confuse people as to the nature of these structures than they are to help.

I've edited the title of the issue to reflect that we really would like better error messages, but not to allow a name of the RHS of a dependent function.

**Comment by davidfstr** [comment]
I personally read `->` more like comma, which is a fairly unopinionated operator, but apparently that's just me.

Better error message it is then. Thanks for looking at this.

**Comment by david-christiansen** [comment]
The analogy of `->` to `,` makes much more sense in a language like Haskell. The Idris equivalent would really be `**`, which also lets you name the left projection and make it accessible in the type of the right projection. Also there, the value that gets a name (on the left side of `**`) is allowed to have a name, and the scope in which it is defined (on the right side of `**`) is not. So the analogy holds here as well!

**Comment by ahmadsalim** [comment]
Any suggestions on what the better parser error message should be like?

**Comment by david-christiansen** [comment]
Perhaps something like "The return type of a dependent function cannot have an explicit name"?
